[Web, keyboard] Locale layout mapping

dkwingsmt opened the pull request:
This PR changes how logical keys are mapped on Web, aligning it with other platforms. Part of https://github.com/flutter/flutter/issues/100456.

This change ensures that non-latin layout can also use letter shortcuts. For example, before this change, on Russian layout, physical key `Ctrl-KeyA` will produce logical key `Ctrl-ф`, which will not trigger any common shortcuts. After the change, it will produce logical key `Ctrl-A`.

### Technical Details

The algorithm is as follows: Analyze every key of the current layout, 
1. If a key can produce an alnum under _some_ modifier, then this key is mapped to this alnum.
2. After the previous step, if some alnum is not mapped, they're mapped to their corresponding key on the US keyboard.
3. The remaining keys are mapped to the unicode plane according to their produced character.

This algorithm, which we call the _benchmark planner_, has been applied to all other desktop platforms.

However, we can't simply apply it to Web: Web DOM API does not tell which keyboard layout the user is on, or how the current layout maps keys (there is a `KeyboardLayout` API that is supported only by Chrome, and explicitly refused by all other browsers). So we have to invent a "blind" algorithm that applies to any layout, while keeping the same result.

Luckily, we're able to fetch a list of "all keyboard layouts" from `Microsoft/VSCode` repo, and we analyzed all layouts beforehand, and managed to combine the result into a huge `code -> key -> result` map. You would imagine it being impossible, since different layouts might use the same `(code, key)` pair for different characters, but in fact such conflicts are surprisingly few, and all the conflicts are mapped to letters. For example, `es-linux` maps `('KeyY', '←')` to `y`, while `de-linux` maps `('KeyY', '←')` to `z`.

We can't distinguished these conflicts only by the `(code, key)` pair, but we can use other information: `keyCode`. Now, `keyCode` is a deprecated property, but we really don't see it being removed any time foreseeable. Also, although `keyCode` is infamous for being platform-dependent, for letter keys it is always equal to the letter character. Therefore such conflicting cases are all mapped to a special value, `0x01`, indicating "use keyCode".

Moreover, to reduce the size of the map, we noticed there are certain patterns that can be easily represented by some `if` statements. These patterns are extracted as the so-called "heuristic mapper". This reduces the map from over 1600 entries to ~450 entries.

To further reduce the package size overhead, this PR also encodes the map into a string that is decoded at run time. This reduces the package size over by ~45% at the cost of code complexity. 

### Structure

A package called `gen_web_locale_keymap` is added to `tools/`, which:
1. Fetches all layouts from `Microsoft/VSCode`
2. Compute the layout map
3. Generate the resulting map data to `web_locale_keymap` (see below)
4. Generate the test cases to `web_locale_keymap`
    - The test cases are an enumeration of all entries (of interest) of all layouts, comparing the result of the resulting mapper class produces the same result as the benchmark planner.

A package called `web_locale_keymap` is added to `third_party`, which does the mapping for all printable characters. Since package must be placed in `third_party` because it uses Microsoft's code (of MIT license).

The SDK rewriter and the license checker are modified to accommodate the new package.

### Package size overhead

Compiling a very simple app that listens to key events:

|   | main.dart.js | Change |
|----|----|----|
| Before PR | 2048730 bytes  | |
| After PR | 2053058 bytes | +4328 bytes (0.21%)|

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

Hixie commented on tools/gen_web_locale_keymap/data/test_cases.dart.tmpl:
```diff
@@ -0,0 +1,30 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
this one should be google iirc

skia-gold commented:
Gold has detected about 1 new digest(s) on patchset 65.
View them at https://flutter-engine-gold.skia.org/cl/github/34625

jonahwilliams commented on web_sdk/sdk_rewriter.dart:
```diff
@@ -11,6 +11,7 @@ final ArgParser argParser = ArgParser()
   ..addOption('output-dir')
   ..addOption('input-dir')
   ..addFlag('ui')
+  ..addFlag('keyboard-layouts')
```
Seems reasonable to me!

mdebbar commented on web_sdk/BUILD.gn:
```diff
@@ -268,7 +288,8 @@ template("_compile_platform") {
          "output_dill must be defined for $target_name")
   _kernel_worker(target_name) {
-    inputs = [ "sdk_rewriter.dart" ] + web_ui_sources
+    inputs =
+        [ "sdk_rewriter.dart" ] + web_ui_sources + web_locale_keymap_sources
```
I'm doing a similar change in [another PR](https://github.com/flutter/engine/pull/37440) and I don't think this line is necessary.

dkwingsmt replied:
I still think this is needed, because `web_ui_sources` only contains all files within `//flutter/lib/web_ui/lib`, while `web_locale_keymap_sources` imports all files within `//flutter/third_party/web_locale_keymap/lib`.

eyebrowsoffire replied:
Actually after reading through this, I think in general the values for `inputs`, both in `_kernel_worker` and in `_dartdevc` are actually sort of bogus. Both of them operate on the already rewritten sources, but are using the pre-rewrite sources as inputs. This kind of works, just because it's sort of transitively capturing the chain of causality. That is to say, the target knows that if `sdk_rewriter.dart` or any of the input files change, the script will be rerun and the outputs of the script might possibly change.

This is kind of janky, however, and what would actually be better is for the `sdk_rewriter` template to accurately reflect each output file that it creates, and have the inputs of the `_kernel_worker` or `_dartdevc` templates be the output of the `sdk_rewriter` template. That's probably out of scope for this change though. I think for now, having `web_locale_keymap_sources` in the inputs is actually probably slightly more correct than not having them. That way, if something changes in `web_locale_keymap_sources` it causes us to rebuild the kernels and JS/sourcemaps.

eyebrowsoffire replied:
I went ahead and filed a follow-on issue here: https://github.com/flutter/flutter/issues/116195

If you are feeling adventurous and want to fix the build rules as part of this PR, have at it. Otherwise, would you mind putting a `TODO(jacksongardner)` with a link to this github issue below it where we are specifying these inputs?

mdebbar commented on web_sdk/BUILD.gn:
```diff
@@ -436,7 +459,8 @@ template("_compile_ddc_modules") {
 
   _dartdevc(target_name) {
-    inputs = [ "sdk_rewriter.dart" ] + web_ui_sources
+    inputs =
+        [ "sdk_rewriter.dart" ] + web_ui_sources + web_locale_keymap_sources
```
Same here.

mdebbar commented on web_sdk/BUILD.gn:
```diff
@@ -585,6 +610,7 @@ if (!is_fuchsia) {
       ":skwasm_impl_library",
       ":skwasm_stub_library",
       ":web_engine_library",
+      ":web_locale_keymap_library",
```
If you want to wait for my [other PR](https://github.com/flutter/engine/pull/37440), you won't need to add this value to many places. There'll be one list at the top of the file. Just add your value there and done:

https://github.com/flutter/engine/blob/5478c2f1457a5235e4bbb7ca250c0f47ad92216f/web_sdk/BUILD.gn#L31-L38

dkwingsmt commented:
@mdebbar I've sync'd the PR with the latest master, addressing as many you comments as possible.

I've made a non-trivial change that compresses the data, which reduces the package size overhead of this PR further by 33%.

I'm still resolving some CI issues. In the mean time, can you take another look?

gspencergoog commented on lib/web_ui/lib/src/engine/keyboard_binding.dart:
```diff
@@ -211,10 +228,28 @@ class FlutterHtmlKeyboardEvent {
+        return locale_keymap.LocaleKeymap.darwin();
+      case OperatingSystem.windows:
+        return locale_keymap.LocaleKeymap.win();
+      default:
```
The `default` case should be avoided: It is better to have cases for each enum entry so that if a new value is added, the linter will highlight this location as one that doesn't handle the new value.

https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-using-if-chains-or--or--with-enum-values

gspencergoog commented on lib/web_ui/lib/src/engine/keyboard_binding.dart:
```diff
@@ -211,10 +228,28 @@ class FlutterHtmlKeyboardEvent {
+      _mapping = _mappingFromPlatform(platform);
 
   final DispatchKeyData performDispatchKeyData;
+  // Whether the current platform is macOS, which affects how certain key events
```
Is it the case that `onMacOS` is *only* used to differentiate things for macOS, or are these keyboard quirks also present on iOS?

dkwingsmt replied:
Nice find. I've tested it on my iPad and I confirm that they also present on iOS. I'm changing it to onDarwin.

dkwingsmt replied:
Let me make it a separate PR after this PR.

gspencergoog commented on lib/web_ui/lib/src/engine/keyboard_binding.dart:
```diff
@@ -273,29 +308,14 @@ class KeyboardConverter {
-      result = result + _kCharLowerA - _kCharUpperA;
-    }
-    return result;
+    return isUpperLetter(key.codeUnitAt(0)) && key.length > 1;
```
Isn't `key.length > 1` sufficient? Or can non-named keys have more than one character? Would it be true that `key.length > 2` would be enough? Just trying to see if we can avoid checking the code unit range entirely.

dkwingsmt replied:
That's a good point. Although I'm not sure if there is a `KeyboardKey.key` that contains multiple JS characters, all keys that `LocaleMapping` handles are single-character, which is all that this function cares about.

dkwingsmt replied:
I think we can remove this range check for now and revise it when we see a counterexample.

dkwingsmt replied:
Just did a quick check of the VSCode layouts. It turns out that `en-in` has a key `L̥` which is two characters.

gspencergoog replied:
Do all named keys have a name that is >2 characters?

dkwingsmt replied:
Pretty sure so. Because all named keys are alnum, therefore having only 1 character conflicts with letter keys.

Also this is the standard for `KeyboardEvent.key`: https://w3c.github.io/uievents-key/ , which clearly says that `key` can consist of a letter and a combining code unit (accents).

dkwingsmt commented:
@gspencergoog I've addressed all your comments. I'd like to add another test for the `en-in` case, but I'm currently having a problem with `felt test`. Nevertheless, can you take another look?

mdebbar reviewed (commented):
LGTM with some minor comments.

If you don't mind, I would like to wait for @eyebrowsoffire's input on the `web_locale_keymap_sources` thing in `web_sdk/BUILD.gn`. If it turns out it's needed, then I have to add `web_unicode_sources` too.

dkwingsmt commented on lib/web_ui/lib/src/engine/keyboard_binding.dart:
```diff
@@ -369,22 +381,33 @@ class KeyboardConverter {
-    final String? character = logicalKeyIsCharacter ? eventKey : null;
-    final int logicalKey = () {
+    final bool logicalKeyIsCharacter = !_eventKeyIsKeyName(eventKey);
+    final ValueGetter<int> logicalKey = _cached<int>(() {
```
This function might not be used: if the event is a key up event, the logical key is simply the pressed logical key, and this function will not be run. In this way, I'm hoping not to run the function body, since it's not a trivial amount of work. And as far as I understand, `late` still always runs the code. (I can add a comment to it.)

mdebbar replied:
According to the docs, the `late` keyword makes the field lazy, which means the right-hand-side won't run until you access the field. At least that's my understanding.

https://dart.dev/null-safety/understanding-null-safety#lazy-initialization

gspencergoog commented on lib/web_ui/lib/src/engine/keyboard_binding.dart:
```diff
@@ -273,29 +304,14 @@ class KeyboardConverter {
+  // Character keys, however, can also have more than 1 code unit: en-in
+  // maps KeyL to L̥/l̥. To resolve this, we check the second code unit.
+  static bool _eventKeyIsKeyName(String key) {
+    return key.length > 1 && key.codeUnitAt(0) < 0x7F && key.codeUnitAt(1) < 0x7F;
```
Can't you just say `key.length > 2`? Are there any names that are only two characters?

dkwingsmt replied:
Yes, such as `Fn` and `F1`. And I'm not sure if there are characters of 3 or more code units.

gspencergoog replied:
Ahh, OK.  LGTM then.

dkwingsmt commented:
@mdebbar I've applied the fix to `BUILD.gn` as you said. Those changes are indeed not necessary.

gspencergoog reviewed (approved):
![32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d](https://user-images.githubusercontent.com/8867023/37737237-66e9c9dc-2d10-11e8-8797-13a405eace49.png)

LGTM, modulo the question about >2 names above.

skia-gold commented:
Gold has detected about 1 new digest(s) on patchset 91.
View them at https://flutter-engine-gold.skia.org/cl/github/34625
